File: src/main/java/de/janitza/maven/gcs/impl/GoogleCloudStorageService.java
package de.janitza.maven.gcs.impl;

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
import com.google.api.client.googleapis.media.MediaHttpUploader;
import com.google.api.client.http.FileContent;
import com.google.api.services.storage.Storage;
import com.google.api.services.storage.model.Bucket;
import com.google.api.services.storage.model.ObjectAccessControl;
import com.google.api.services.storage.model.StorageObject;
import de.janitza.maven.gcs.api.IGoogleCloudStorageService;
import de.janitza.maven.gcs.api.config.IGCSConfig;
import de.janitza.maven.gcs.api.config.loader.IServiceAccountCredentials;
import de.janitza.maven.gcs.impl.utils.HttpUtil;
import de.janitza.maven.gcs.impl.utils.Lists;
import de.janitza.maven.gcs.impl.utils.StoragePath;
import org.apache.maven.plugin.logging.Log;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Created with IntelliJ IDEA.
 * User: Jan Müller <jan.mueller@janitza.de>
 * Date: 10.02.15
 * Time: 11:48
 */
public class GoogleCloudStorageService implements IGoogleCloudStorageService {


    public static final String ROLE_READER = "READER";
    public static final String USER_ALL_USERS = "allUsers";

    private final IGCSConfig m_IGCSConfig;
    private final GoogleCredential m_Credentials;
    private final Storage m_Client;
    private final Bucket m_Bucket;
    private final Log m_Log;

    public GoogleCloudStorageService(
            final IGCSConfig igcsConfig,
            final Log log
    ) throws IOException {
        m_Log = log;
        m_IGCSConfig = igcsConfig;
        m_Credentials = authorize();
        m_Client = new Storage.Builder(
                m_IGCSConfig.getHttpTransport(),
                m_IGCSConfig.getJsonFactory(),
                m_Credentials
        ).setApplicationName(
                m_IGCSConfig.getGCSApplicationName()
        ).build();
        m_Bucket = m_Client.buckets()
                .get(m_IGCSConfig.getBucketName())
                .setProjection("full")
                .execute();
        logBucketInfo(m_IGCSConfig.getBucketName(), m_Bucket);
    }

    private GoogleCredential authorize() {
        final IServiceAccountCredentials serviceAccountCredentials =
                m_IGCSConfig.getServiceAccountCredentials();

        return new GoogleCredential.Builder()
                .setTransport(m_IGCSConfig.getHttpTransport())
                .setJsonFactory(m_IGCSConfig.getJsonFactory())
                .setServiceAccountScopes(m_IGCSConfig.getScopes())
                .setServiceAccountId(serviceAccountCredentials.getAccountId())
                .setServiceAccountPrivateKey(serviceAccountCredentials.getPrivateKey())
                .build();
    }

    @Override
    public void uploadFile(
            final Path file,
            final Optional<String> relativePathInStorage,
            boolean sharePublic
    ) throws IOException {

        final String fileName = file.getFileName().toString();
        final String storagePath = getStoragePath(relativePathInStorage, fileName);

        final StorageObject object = createStorageObject(
                storagePath,
                file,
                sharePublic,
                m_Bucket
        );

        logFileUploading(file, storagePath);

        final Instant t1 = Instant.now();

        retry(file, object, 10);

        logFileUploaded(file, storagePath, t1);
    }

    private Storage.Objects.Insert createInsert(
            final Path file,
            final StorageObject object
    ) throws IOException {
        final long fileSize = Files.size(file);
        final Storage.Objects.Insert insertion = m_Client.objects().insert(
                m_Bucket.getName(),
                object,
                new FileContent(object.getContentType(), file.toFile())
        );

        final MediaHttpUploader mediaHttpUploaderInsertion = insertion
                .getMediaHttpUploader();
        if (mediaHttpUploaderInsertion != null) {
            mediaHttpUploaderInsertion.setDirectUploadEnabled(false);
            mediaHttpUploaderInsertion
                    .setProgressListener(new GCSProgressListener(fileSize, m_Log));
        }
        return insertion;
    }

    private void retry(
            final Path file,
            final StorageObject object,
            final int maxRetryCount
    ) throws IOException {
        final IOException exception = new IOException("Retry failed!");
        int i = 0;
        while (i < maxRetryCount) {
            try {
                createInsert(file, object).execute();
                break;
            } catch (IOException e) {
                exception.addSuppressed(e);
                i++;
                if (i < maxRetryCount) {
                    m_Log.info(
                            "Upload was interrupted. Retrying the upload! " +
                                    "Resuming is currently not supported by the" +
                                    " google lib!"
                    );
                }
            }
        }
        if (i == maxRetryCount) {
            throw exception;
        }
    }

    private void logBucketInfo(
            final String bucketName,
            final Bucket bucket
    ) {
        m_Log.info("Bucket name: " + bucketName);
        m_Log.info("Bucket location: " + bucket.getLocation());
    }

    private void logFileUploaded(
            final Path file,
            final String storagePath,
            final Instant t1
    ) {
        m_Log.info(
                String.format(
                        "Uploaded %s to %s in %s",
                        file,
                        storagePath,
                        Duration.between(t1, Instant.now())
                )
        );
    }

    private void logFileUploading(
            final Path file,
            final String storagePath
    ) {
        m_Log.info(
                String.format(
                        "Uploading %s to %s",
                        file,
                        storagePath
                )
        );
    }

    private String getStoragePath(
            final Optional<String> relativePathInStorage,
            final String fileName
    ) {
        return relativePathInStorage.map(
                (rp) -> rp.isEmpty() ? fileName : StoragePath.join(rp, fileName)
        ).orElse(fileName);
    }

    private StorageObject createStorageObject(
            final String storagePath,
            final Path file,
            final boolean sharePublic,
            final Bucket bucket
    ) throws IOException {
        final String fileName = file.getFileName().toString();
        StorageObject object =
                new StorageObject()
                        .setName(storagePath)
                        .setContentDisposition(
                                HttpUtil.getContentDisposition(fileName))
                        .setContentType(HttpUtil.getMimeType(file))
                        .setSize(size(file));

        if (sharePublic) {
            object.setAcl(addPublicReadAccess(getDefaultObjectAcl(bucket)));
        }

        return object;
    }

    private BigInteger size(final Path file) throws IOException {
        return BigInteger.valueOf(Files.size(file));
    }

    private List<ObjectAccessControl> addPublicReadAccess(
            final List<ObjectAccessControl> defaultAcl
    ) {
        final List<ObjectAccessControl> acl = new ArrayList<>(defaultAcl);

        final boolean alreadyShared = !acl.stream().anyMatch((oac) -> {
            return USER_ALL_USERS.equals(oac.getEntity());
        });

        if (alreadyShared) {
            acl.add(
                    new ObjectAccessControl()
                            .setEntity(USER_ALL_USERS)
                            .setRole(ROLE_READER)
            );
        }

        return acl;
    }

    private List<ObjectAccessControl> getDefaultObjectAcl(final Bucket bucket) {
        return Lists.ensureAList(bucket.getDefaultObjectAcl());
    }
}

File: src/main/scala/de/janitza/maven/gcs/impl/GoogleCloudStorageService.scala
package de.janitza.maven.gcs.impl

import java.io.IOException
import java.math.BigInteger
import java.nio.file.{Files, Path}
import java.time.{Duration, Instant}

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
import com.google.api.client.http.FileContent
import com.google.api.services.storage.Storage
import com.google.api.services.storage.model.{Bucket, ObjectAccessControl, StorageObject}
import de.janitza.maven.gcs.api.config.GCSConfig
import de.janitza.maven.gcs.api.{Error, IGoogleCloudStorageService, Result, Success}
import de.janitza.maven.gcs.impl.util.{HttpUtil, StoragePath}
import org.apache.maven.plugin.logging.Log

import scala.collection.JavaConverters._


object GoogleCloudStorageService {
  val ROLE_READER = "READER"
  val USER_ALL_USERS = "allUsers"
  val PROJECTION = "full"
  val MAX_RETRY_ATTEMPTS = 10
}

class GoogleCloudStorageService @throws[IOException]
(val gcsConfig: GCSConfig, val log: Log)
  extends IGoogleCloudStorageService {

  import GoogleCloudStorageService._

  private val m_Client =
    new Storage.Builder(gcsConfig.httpTransport, gcsConfig.jsonFactory, authorize)
      .setApplicationName(gcsConfig.gcsApplicationName)
      .build
  private val m_Bucket =
    m_Client.buckets
      .get(gcsConfig.bucketName)
      .setProjection(PROJECTION)
      .execute

  logBucketInfo(gcsConfig.bucketName, m_Bucket)

  private def authorize: GoogleCredential = {
    val serviceAccountCredentials = gcsConfig.serviceAccountCredentials
    new GoogleCredential.Builder()
      .setTransport(gcsConfig.httpTransport)
      .setJsonFactory(gcsConfig.jsonFactory)
      .setServiceAccountScopes(gcsConfig.scopes.asJava)
      .setServiceAccountId(serviceAccountCredentials.accountId)
      .setServiceAccountPrivateKey(serviceAccountCredentials.privateKey)
      .build
  }

  override def uploadFile(file: Path, relativePathInStorage: Option[String], sharePublic: Boolean): Result[Unit] = {
    val storagePath = getStoragePath(relativePathInStorage, file.getFileName.toString)
    val storageObjectResult = createStorageObject(storagePath, file, sharePublic, m_Bucket)
    storageObjectResult match {
      case Success(storageObject) => Success(uploadFile(file, storagePath, storageObject))
      case e: Error => e
    }
  }

  def uploadFile(file: Path, storagePath: String, storageObject: StorageObject): Unit = {
    logFileUploading(file, storagePath)
    val t1 = Instant.now
    insertWithRetry(file, storageObject, MAX_RETRY_ATTEMPTS)
    logFileUploaded(file, storagePath, t1)
  }

  @throws[IOException]
  private def createInsert(file: Path, storageObject: StorageObject): Storage#Objects#Insert = {
    val fileSize: Long = Files.size(file)
    val insertion: Storage#Objects#Insert =
      m_Client.objects.insert(
        m_Bucket.getName,
        storageObject,
        new FileContent(storageObject.getContentType, file.toFile)
      )
    val mediaHttpUploaderInsertion = insertion.getMediaHttpUploader
    if (mediaHttpUploaderInsertion != null) {
      mediaHttpUploaderInsertion.setDirectUploadEnabled(false)
      mediaHttpUploaderInsertion.setProgressListener(new GCSProgressListener(fileSize, log))
    }
    insertion
  }

  private def insertWithRetry(file: Path, storageObject: StorageObject, maxRetryCount: Int): Result[Unit] = {
    val errorResults = Stream.from(1).take(maxRetryCount)
      .map(Insertion(file, storageObject).execute)
      .takeWhile(result => result.isInstanceOf[Error]).toSeq
    if (errorResults.nonEmpty && errorResults.size < maxRetryCount) {
      errorResults.last
    } else {
      Success()
    }
  }

  private case class Insertion(file: Path, storageObject: StorageObject) {
    def execute(counter: Int): Result[Unit] = {
      try {
        createInsert(file, storageObject).execute
        Success()
      } catch {
        case e: IOException => {
          val errorMessage: String = "Upload was interrupted."
          log.info(errorMessage)
          Error(errorMessage, Some(e))
        }
      }
    }
  }

  private def logBucketInfo(bucketName: String, bucket: Bucket) {
    log.info(s"Bucket name: $bucketName")
    log.info(s"Bucket location: ${bucket.getLocation}")
  }

  private def logFileUploaded(file: Path, storagePath: String, t1: Instant) {
    log.info(s"Uploaded $file to $storagePath in ${Duration.between(t1, Instant.now)}")
  }

  private def logFileUploading(file: Path, storagePath: String) {
    log.info(s"Uploading $file to $storagePath")
  }

  private def getStoragePath(relativePathInStorage: Option[String], fileName: String) =
    relativePathInStorage.map({
      case rp: String if rp.isEmpty => fileName
      case rp: String => StoragePath.join(rp, fileName)
    }) getOrElse fileName

  private def createStorageObject(storagePath: String, file: Path, sharePublic: Boolean, bucket: Bucket): Result[StorageObject] = {
    val fileName = file.getFileName.toString
    HttpUtil.getMimeType(file) match {
      case Success(mimeType) => size(file) match {
        case Success(fileSize) =>
          Success(createStorageObject(storagePath, fileName, fileSize, sharePublic, bucket, mimeType))
        case Error(message, exception) => Error(message, exception)
      }
      case Error(message, exception) => Error(message, exception)
    }
  }

  private def createStorageObject(storagePath: String, fileName: String, fileSize: BigInteger, sharePublic: Boolean, bucket: Bucket, mimeType: String): StorageObject = {
    val storageObject = new StorageObject()
      .setName(storagePath)
      .setContentDisposition(HttpUtil.getContentDisposition(fileName))
      .setContentType(mimeType)
      .setSize(fileSize)
    if (sharePublic) {
      storageObject.setAcl(addPublicReadAccess(getDefaultObjectAcl(bucket)).asJava)
    }
    storageObject
  }

  @throws[IOException]
  private def size(file: Path): Result[BigInteger] = {
    try {
      Success(BigInteger.valueOf(Files.size(file)))
    } catch {
      case e: IOException => Error("Couldn't determine size of file to be uploaded!", Some(e))
    }
  }

  private def addPublicReadAccess(defaultAcl: Seq[ObjectAccessControl]): Seq[ObjectAccessControl] = {
    val alreadyShared = defaultAcl.toStream.map(_.getEntity).exists(USER_ALL_USERS.equals(_))
    if (alreadyShared)
      defaultAcl
    else
      defaultAcl :+
        new ObjectAccessControl()
          .setEntity(USER_ALL_USERS)
          .setRole(ROLE_READER)
  }

  private def getDefaultObjectAcl(bucket: Bucket) =
    Option(bucket.getDefaultObjectAcl).map {
      _.asScala.toList
    } getOrElse Seq()
}

